dace/codegen/targets/intel_fpga.py

```diff
@@ -524,26 +542,51 @@ def generate_kernel_internal(self, sdfg, state, kernel_name, subgraphs,
         kernel_stream.write(kernel_header_stream.getvalue() +
                             kernel_body_stream.getvalue())
 
-        self.generate_host_function_epilogue(sdfg, state, host_code_body_stream)
+        # Generate host kernel invocation
+        self.generate_host_function_body(sdfg, state, state_host_body_stream,
+                                         kernel_name, predecessors)
 
-        # Store code to be passed to compilation phase
-        self._host_codes.append(
-            (kernel_name, host_code_header_stream.getvalue() +
-             host_code_body_stream.getvalue()))
+    def generate_host_function_prologue(self, sdfg, state, host_stream,
+                                        kernel_name):
+        seperator = "/" * 59
+        host_stream.write(
+            f"\n{seperator}\n// Kernel: {kernel_name}\n{seperator}\n\n")
 
-    @staticmethod
-    def generate_host_function_prologue(sdfg, state, host_stream):
-        host_stream.write("std::vector<hlslib::ocl::Kernel> kernels;", sdfg,
-                          sdfg.node_id(state))
+        host_stream.write(
+            f"std::vector<hlslib::ocl::Kernel> {kernel_name}_kernels;", sdfg,
+            sdfg.node_id(state))
 
-    @staticmethod
-    def generate_host_function_epilogue(sdfg, state, host_stream):
+    def generate_host_function_body(self, sdfg: dace.SDFG,
+                                    state: dace.SDFGState,
+                                    host_stream: CodeIOStream, kernel_name: str,
+                                    predecessors: list):
+        '''
+        Generate the host-specific code for spawning and synchronizing the given kernel.
+        :param sdfg:
+        :param state:
+        :param host_stream: Device-specific code stream
+        :param kernel_name:
+        :param predecessors: list containing all the name of kernels that must be finished before starting this one
+        '''
         state_id = sdfg.node_id(state)
-        host_stream.write(
-            "const auto start = std::chrono::high_resolution_clock::now();",
-            sdfg, state_id)
         launch_async = Config.get_bool("compiler", "intel_fpga", "launch_async")
+
+        # Check if this kernel depends from other kernels
+        needs_synch = len(predecessors) > 0
+
+        if needs_synch:
+            # Build a vector containing all the events associated with the kernels from which this one depends
+            kernel_deps_name = f"deps_{kernel_name}"
+            host_stream.write(f"std::vector<cl::Event > {kernel_deps_name};")
+            for pred in predecessors:
+                # concatenate events from predecessor kernel
+                host_stream.write(
+                    f"{kernel_deps_name}.insert({kernel_deps_name}.end(), {pred}_events.begin(), {pred}_events.end());"
```

Comment:
I think we should make this a constant size array instead of a vector, creating and appending a vector only creates unnecessary overhead when we anyway know the size at compile-time.